zip-0316.rst

```diff
@@ -604,15 +612,53 @@ are reserved to indicate Metadata Items other than Receivers or
 Viewing Keys. These Items MAY affect the overall interpretation of
 the UA / UVK (for example, by specifying an expiration date).
 
+.. _`MUST-understand typecodes`:
+
+As of `Revision 1`_, the subset of Metadata typecodes in
+the range :math:`mathtt{0xE0}` to :math:`mathtt{0xEF}` are designated
+as "MUST-understand": if a Consumer is unable to recognize the meaning
+of a Metadata Item with a typecode in this range, then it MUST regard
+the entire UA/UVK as unsupported and not process it further.
+
 Since Metadata Items are not Receivers, they MUST NOT be selected by
 a Sender when choosing a Receiver to send to, and since they are not
 Viewing Keys, they MUST NOT provide additional authority to view
 information about transactions.
 
-Currently no Metadata Types are defined. New Metadata Types SHOULD be
-introduced either by a modification to this ZIP or by a new ZIP, in
-accordance with the ZIP Process [#zip-0000]_.
-
+Address Expiration Metadata
+===========================
+
+As of `Revision 1`_, typecodes :math:`\mathtt{0xE0}` and
+:math:`\mathtt{0xE1}` are reserved for address expiry metadata.
+
+The value of a :math:`\mathtt{0xE0}` item MUST be an unsigned 32-bit integer in
+little-endian order specifying the Address Expiry Height, a block height of the
+Zcash chain associated with the UA/UVK. A Unified Address containing metadata
+typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
+the Zcash chain is greater than this value.
+
+The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
+little-endian order specifying a Unix Epoch Time, hereafter referred to as the
+Address Expiry Time. A Unified Address containing metadata typecode
+:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
+after the Address Expiry Time.
+
+A Sender that supports `Revision 1`_ of this specification MUST set
+a non-zero ``nExpiryHeight`` field in transactions that it creates. If the
+``nExpiryHeight`` normally constructed by the Sender would be greater than the
+Address Expiry Height, then the transaction MUST NOT be sent. If only an
+Address Expiry Time is specified, then the Sender MUST choose a value for
+``nExpiryHeight`` such that the transaction will expire no more than 24 hours
+after the current time.
+
+When deriving from a UFVK or UIVK containing typecodes :math:`\mathtt{0xE0}`
+and/or :math:`\mathtt{0xE1}`, these Metadata Items MUST be retained unmodified
+in the output.
+
+The intent of this specification is that Consumers of Unified Addresses MUST
+NOT send to expired addresses. If only an Address Expiry Time is specified, a
+transaction to the associated address could be mined after the Address Expiry
+Time within a 24-hour window.
 
```

Comment:
Add a Rationale subsection explaining:
* That the reason to specify that the transaction MUST NOT be sent when its `nExpiryHeight` as normally constructed is greater than the Address Expiry Height &mdash;rather than limiting the `nExpiryHeight` field&mdash;is to avoid unnecessary information leakage in that field about which address was used as the destination.
* The intent behind requiring the Sender to choose a transaction that expires no more than 24 hours after the current time (i.e. that an adversary could reintroduce the tx an arbitrary time later, if it is still valid).

Also say that if a wallet user attempts to send to an expired address, the error presented to the user SHOULD include a suggestion to attempt to find a more recent address for the intended recipient.